### ISSUE TITLE
Introduce process rate DDT for PUMAS #589

### ISSUE THREAD

**Issue by cacraigucar** [ESCOMP/CAM#589]
Create a DDT which will contain the process rate variables for PUMAS

**Comment by cacraigucar** [comment]
Notes from a joint design session with @cacraigucar  and @gold2718 

micro_pumas_diags.F90 code snippet:

```
   type, public :: proc_rates_type
      real (r8),allocatable :: prodsnow(:,:)
   contains
      procedure :: allocate => proc_rates_allocate
      procedure :: deallocate => proc_rates_deallocate
   end type proc_rates_type

contains

   subroutine proc_rates_allocate (psetcols, nlev)
      allocate(prodsnow(psetcols,nlev),stat=ierr)
   end subroutine proc_rate_allocate


```
----------------

code snippet for micro_pumas_cam.F90:

```
   type(proc_rates_type :: proc_rates


   call proc_rates%allocate(psetcols,nlev)

   call proc_rates%deallocate(psetcols,nlev)

```
Notes:

   add_hist_coord('trop_cld_nlev', nlev,...  <--- This might go into ref_pres.F90 and values are top_lev to pver
     units 1 , long_name= "model level number" values=[top_lev, top_lev+1...,pver]

   EXAMPLE :cospsimulator_intr.F90 line 622

   Use trop_cld_nlev in addfld call for the elements in the new DDT


   REQUIRED CHANGE:  (this allows pbuf to still be pver, but allow nlev for PUMAS)

```
  if (qcsevap_idx > 0 ) qcsevapout_grid_ptr  = qcsevapout_grid
```
  to
```
  if (qcsevap_idx > 0 ) qcsevapout_grid_ptr(:,top_lev:)  = qcsevapout_grid
```
@nusbaume - Can the ADF handle the output being nlev instead of pver?
@andrewgettelman - Do you see any issues with the above code snippets?  Let us know if you need more details.

**Comment by nusbaume** [comment]
The ADF should be able to handle it, although it might need a few very minor modifications to make sure the ADF is looking for the `nlev` dimension name in the relevant history files.

**Comment by andrewgettelman** [comment]
1. Why should the sizes of the arrays (vertical dimension) be any different than they are now for things in the calling tree? I think they should stay the same for simplicity sake. I do not think this should add a special dimension to the history file. Seems to add complication which is not there now. The user of output shouldn't really see a change. 

2. I'm not totally following where this code would go and how it would look to the user, or how one would add something. I guess it would copy the prodsnow variable, but i'm not sure how this would look from what you have shown.

**Comment by gold2718** [comment]
> 1. Why should the sizes of the arrays (vertical dimension) be any different than they are now for things in the calling tree? I think they should stay the same for simplicity sake.

If these items are in a DDT and that DDT is passed to PUMAS, the arrays inside have to be the correct size (`nlev` not `pver`). There is no way to create `pver`-sized arrays inside the DDT but only pass in `nlev` levels or each array.

**Comment by gold2718** [comment]
> 2\. I'm not totally following where this code would go and how it would look to the user, or how one would add something. I guess it would copy the prodsnow variable, but i'm not sure how this would look from what you have shown.

I think most of this information should be in a PUMAS issue because that is where the new (diagnostic DDT) module will live.
CAM will allocate an instance of the DDT, pass it into PUMAS, and call `outfld` on fields in the DDT after the call. We could copy the `nlev` arrays into `pver` arrays to call `outfld` for full height (more disk space) fields on file but why go through all that trouble? One call at startup to `add_hist_coord` will allow output of the arrays as computed by PUMAS.

**Comment by cacraigucar** [comment]
An email from Andrew indicated these are the fields to initially include in the DDT:
prodsnow, evapsnow
qcsevap,            qisevap,            qvres,              &
     cmeitot,            vtrmc,              vtrmi,              &
     umr,                          ums,                          &
     umg,                qgsedten,                               &
     qcsedten,                     qisedten,                     &
     qrsedten,                     qssedten,                     &
     pratot,                       prctot,                       &
     mnuccctot,          mnuccttot,          msacwitot,          &
     psacwstot,          bergstot,  vapdepstot,         bergtot,            &
     melttot,    meltstot,   meltgtot,       homotot,            &
     qcrestot,           prcitot,            praitot,            &
     qirestot,           mnuccrtot,    mnudeptot,       mnuccritot, pracstot,           &
     meltsdttot,         frzrdttot,          mnuccdtot,          &
     pracgtot,           psacwgtot,          pgsacwtot,          &
     pgracstot,          prdgtot,           &
     qmultgtot,          qmultrgtot,         psacrtot,           &
     npracgtot,          nscngtot,           ngracstot,          &
     nmultgtot,          nmultrgtot,         npsacwgtot

**Comment by cacraigucar** [comment]
After side discussions on the feasibility of having the DDT variables output on a smaller grid versus copying to the full grid so they could be output @andrewgettelman concluded that we will proceed with outputting them on the smaller grid.  His concern was that a copy to the larger grid would slow down GPU adoption, even if the copy were in the interface layer.

**Comment by cacraigucar** [comment]
@andrewgettelman and @adamrher  - Is there any reason to keep micro_mg_tend1_0 in cam_dev and the version of PUMAS which supports it?  This DDT will need to be added to the 1_0 version as well, and I wanted to make sure that it is still a version that is wanted before I make the edits to it as well.

**Comment by andrewgettelman** [comment]
I do not see a reason to keep micro_mg_tend1_0 in cam_dev and PUMAS. We do not need to add the DDT to v1_0, and we should avoid it if we can. So I guess we can remove micto_mg_tend1_0 from the latest pumas on cam_dev. Hopefully.

**Comment by gold2718** [comment]
@andrewgettelman, Is there a reason some `micro_pumas_tend` `intent(out)` variables (e.g., `qcsinksum_rate1ord`) are not slated to be added to the output DDT?

**Comment by andrewgettelman** [comment]
The original intent was to be diagnostics only for the DDT. the qcsinksum_rate1ord is used by aerosols, and is not really a diagnostic (and not something I really understand).

**Comment by cacraigucar** [comment]
@andrewgettelman - I have just stumbled across the variable bergstot from your list (known as bergso in the CAM interface code) which is a member of the pbuf.  I was just starting to puzzle out how to handle this, but from your comment it sounds like it may not need to be a member of the DDT.  Is that correct?

Also, I will have this same issue with a number of variables that are pbuf members when SILHS is turned on.  Right now these are members of the DDT, but I will need to figure out to handle them for the SILHS case.  An example of this is qcsedten, but there are a handful of these variables which can be seen in micro_pumas_cam_register at the bottom of the routine in the fields for "SILHS hole filling"

**Comment by andrewgettelman** [comment]
Hmm... these actually are part of the budget terms we need. I would prefer if bergstot/bergso and possibly some of the fields needed by SILHS be put in the DDT and the CAM interface can pull them out to pbuf simply I assume?

**Comment by cacraigucar** [comment]
An email from @andrewgettelman indicated that these additional fields should be added to the DDT:

nnuccctot,          nnuccttot,          nnuccdtot,          &
     nnudeptot,          nhomotot,           nnuccrtot,          &
     nnuccritot,         nsacwitot,          npratot,            &
     npsacwstot,         npraitot,           npracstot,          &
     nprctot,            nprcitot,           ncsedten,           &
     nisedten,           nrsedten,           nssedten,           &
     ngsedten,           nmelttot,           nmeltstot,          &
     nmeltgtot

**Comment by nusbaume** [comment]
This issue should have been closed by #632.  Closing now.
